cv32e40s/tests/programs/custom/xsecure_test/xsecure_test.c

```diff
@@ -96,97 +182,194 @@ void u_sw_irq_handler(void) {
-  __asm__ volatile("csrrw %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrs %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrc %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(csr_read));
+  // read ctrlcpu and secureseed in user mode (which should not be successful):
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
+  __asm__ volatile("csrrs %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
+  __asm__ volatile("csrrc %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
 
-  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrs %0, " STR(CSRADDR_SECURESEED0) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrc %0, " STR(CSRADDR_SECURESEED0) ", x0" : "=r"(csr_read));
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
 
-  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED1) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrs %0, " STR(CSRADDR_SECURESEED1) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrc %0, " STR(CSRADDR_SECURESEED1) ", x0" : "=r"(csr_read));
+  // number of exceptions should equal number of acesses
+  assert_or_die(num_trap_executions, 3+3, "error: accessing cpuctrl or secureseed_ in usermode does not cause a trap\n");
+
+  // resetting the trap handler count
+  num_trap_executions = 0;
+
+  // enter machine mode:
+  set_m_mode();
+
+  // read ctrlcpu and secureseed in machine mode (which should be successful):
 
-  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED2) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrs %0, " STR(CSRADDR_SECURESEED2) ", x0" : "=r"(csr_read));
-  __asm__ volatile("csrrc %0, " STR(CSRADDR_SECURESEED2) ", x0" : "=r"(csr_read));
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
+  __asm__ volatile("csrrs %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
+  __asm__ volatile("csrrc %0, " STR(CSRADDR_CPUCTRL) ", x0" : "=r"(rd));
+
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
+
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
+
+  __asm__ volatile("csrrw %0, " STR(CSRADDR_SECURESEED0) ", %1" : "=r"(rd) : "r"(rand()));
 
   // number of exceptions should equal number of acesses
-  assert_or_die(num_trao_executions, 4*3, "error: accessing cpuctrl or secureseed_ in usermode does not cause a trap\n");
+  assert_or_die(num_trap_executions, 0, "error: accessing cpuctrl or secureseed_ in machine mode cause a trap\n");
 
   // set the trap handler to go into default mode
   trap_handler_beh = UNEXPECTED_IRQ_BEH;
 
 }
 
 
+
 int main(void){
 
   unexpected_irq_beh = 0;
 
+  // seed the rand() function, which is used in some of the test functions
```

Comment:
Please use a source of random from the systemverilog environment (have a look at interrupt_test.c, `random_num(...)` on how to do that. 

As you seed `rand()` with `srand(1)`, this test will always use the same fixed (but random) value. Furthermore, rand() is not guaranteed to be stable across different simulators, versions and systems, so this can cause issues when trying to reproduce a specific failure